### ISSUE TITLE
Unable to expand admin tree view items in Firefox under some conditions

### ISSUE THREAD

**Issue by damien-git** [blueriver/MuraCMS#2672]
We have a new issue with the admin UI: tree view items do not expand. The XHR returns nothing.

This happens with Firefox but not Chrome or Edge, and it does not happen with all our Mura instances, so IIS and ColdFusion settings could be different. When it happens, Firefox sends 2 requests instead of 1. The second request returns nothing. The same thing happens with all browsers if the cArch.loadSiteSection request is sent twice (nothing is returned). It is not clear to me at this point why Firefox sends 2 requests. The versions of Firefox we have tried are 52.4.0 (ESR) and 56.0.1. I suspect an issue with the headers, but Firefox network monitor only shows 1 request, so it's hard to tell what headers it is getting.

**Comment by damien-git** [comment]
This issue only happens with SSL.

**Comment by mattlevine** [comment]
Does this happen while trying to expand all nodes? What happens when you logout and back in again?

-Matt

**Comment by damien-git** [comment]
It happens whenever I try to expand any node. Nothing is different if I log out and log in. We were able to reproduce the behavior on several computers.

**Comment by mattlevine** [comment]
Hmm, I can't seem to reproduce this.  I have a testing instance with what I think is a pretty similar stack, include SSL and Shibboleth?

What version of FF?

**Comment by damien-git** [comment]
Yes, the problematic server has SSL and Shibboleth, but then we have another server with SSL and Shibboleth that does not exhibit the issue. However the problem goes away if we disable SSL.
As I wrote earlier, we tried the latest versions of Firefox (56.0.1) and Firefox ESR (52.4.0).
I have tried to make the response headers the same between our 2 servers, to try to find out in what exact circumstance Firefox asks for another request. At this point the only difference I see is that the working server sends a Set-Cookie parameter that the other does not send.

**Comment by damien-git** [comment]
It looks like there are syntax errors in the JSON data (escaped quotes), but it is the same between our 2 servers, so I am guessing it is not causing the issue.

**Comment by mattlevine** [comment]
Interesting, could there be something about the content that is exposing the issue?  Like some sort of unicode character?

**Comment by damien-git** [comment]
I didn't see any obvious unicode character, and the content returned is the same between our 2 servers (except for the IDs).

**Comment by mattlevine** [comment]
You said there are two calls happening on the server that is now working correctly?  Is first one being redirected?

**Comment by damien-git** [comment]
no, there is nothing returned the second time when the same URL is called twice (no matter the browser or the server). And apparently this is what is happening when the issue occurs (the browser sends the request twice, and ignores the first one).

**Comment by mattlevine** [comment]
Do you seen the same behavior on other environments?

**Comment by damien-git** [comment]
Other client-side environments, yes. We have only tested 2 server environments.

**Comment by damien-git** [comment]
Looking at the request headers, I see one difference between the 2 requests made by Firefox: The first one has connection: close, while the second one has connection: keep-alive. It is the only difference.

**Comment by damien-git** [comment]
I have been looking at the requests in onRequestStart_include.cfm. This is how I noticed Firefox was sending 2 requests.

**Comment by mattlevine** [comment]
The problem is that since FireFox is sending two requests the second one is closing the section.  I;m not sure why this has suddenly started happening on FF, Mura's code that implements this has basically not changed for many years.  I think I may have reproduced it.  I'll see if I can work around the FF bug.

**Comment by damien-git** [comment]
Thanks ! I wish I could figure out why Firefox does this, but a workaround would be useful anyway, we never know how browser behavior will change in the future.

**Comment by mattlevine** [comment]
As a test can you manually update this file:

https://github.com/blueriver/MuraCMS/blob/develop/admin/core/views/carch/loadsitesection.cfm

and then logout and back in again and retest.

**Comment by damien-git** [comment]
It works, thanks !

**Comment by damien-git** [comment]
Strangely it also expands the last item in the folder if it can be expanded (with all browsers)...

**Comment by mattlevine** [comment]
Actually, that "fix" seems be making things worse.

**Comment by mattlevine** [comment]
Please manually update this file one more time.

https://github.com/blueriver/MuraCMS/blob/develop/admin/core/views/carch/loadsitesection.cfm

**Comment by damien-git** [comment]
This new version is not working for me, it just behaves like it did before. I am not seeing any error in the CF log. I have only changed loadsitesection.cfm, should I update other files ?

**Comment by gowestweb** [comment]
Hi Damien. 
I suspect you are on a windows server, with IIS10 , is that correct? 
If so this is due to the new HTTP/2 protocol in IIS 10.

JC Gomez explains it pretty well here: 
http://www.giancarlogomez.com/2016/05/coldfusion-iis-10-http2-safari-bug.html
except in this case you are seeing a second 'downgraded' request, rather than the endless loop he describes.  

In our own debugging and log analysis, we were able to see that the first request was using HTTP/2, and the second one HTTP/1.1, every time.

We were able to replicate the issue on a new IIS10 server w/ CF 2016 at GoWest Hosting, and the Firefox browser on Matt's local machine. 

You'll see JC also offers simple directions for turning off HTTP/2, by making a few entries in the Windows registry. (You can find these same suggestions elsewhere, if you search for "disable http/2 on IIS 10".)

I made the suggested registry edits on the server, and the problem was resolved. 
It is of course up to you whether to turn off http/2 in production in order to accommodate the fringe behavior in this case, but I believe that will be the remedy for the problem with the double http requests in your Mura admin "tree view".

**Comment by damien-git** [comment]
Yes, we are using IIS10 in the server with the issue. We are going to try to disable HTTP/2 tonight.
From what I understand, the Tomcat versions shipping with CF11 (which we use) and CF2016 today do not support HTTP/2, so I am guessing disabling HTTP/2 should not decrease performance significantly (and might improve it when the double request issue is avoided).

**Comment by gowestweb** [comment]
Great, let us know the results, thank you.

**Comment by damien-git** [comment]
Disabling HTTP/2 in IIS resolved the issue (I reverted to the old loadsitesection.cfm). It also resolved that worse issue with some versions of Safari (which could not load any page or could just load the home page), and it also resolved yet another problem: the HTTP status codes were not being set correctly in Firefox. So this is probably the best solution as of now.
Thanks !
